### PR TITLE
Reformat all OCaml code with ocamlformat

### PR DESCRIPTION
**Pull request by mjambon**
It uses the default settings but we could customize them if necessary. We could also change the settings later at anytime.

I reported a [trailing whitespace bug](https://github.com/ocaml-ppx/ocamlformat/issues/1656) along the way, which required me to move 2 comments. Everything else looked good to me.

I still need to take care of installing ocamlformat as a dependency.

PR checklist:
- [x] changelog is up to date

### CODE REVIEW

File: spacegrep/src/lib/Print.ml [hunk 1]
```diff
@@ -59,8 +55,7 @@ module Debug = struct
   let show_loc = false
 
   let print_loc buf (start, _end_) =
-    if show_loc then
-      bprintf buf "%3i: " start.Lexing.pos_lnum
+    if show_loc then bprintf buf "%3i: " start.Lexing.pos_lnum
```
[the remark this answers is not in the document]
Comment (mjambon) [reply spacegrep/src/lib/Print.ml:58]: Um, there's no else. We need to fix this.

File: toy-matcher/Matcher.ml [hunk 2]
```diff
@@ -364,17 +310,13 @@ module Cache_key = struct
   type t = ellipsis * env * pattern * input
 
   let equal (ellipsis1, env1, pat1, input1) (ellipsis2, env2, pat2, input2) =
-    pat_equal pat1 pat2
-    && (==) input1 input2
-    && ellipsis1 = ellipsis2
-    && Bindings.equal (=) env1.min_env env2.min_env
+    pat_equal pat1 pat2 && input1 == input2 && ellipsis1 = ellipsis2
```
[the remark this answers is not in the document]
Comment (mjambon) [reply toy-matcher/Matcher.ml]: It's getting really hard to use physical equality.

**Comment by mjambon** [comment]
We're getting errors:

```
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   | Id
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   | xs
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   | ")"
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   | ")"
{"matches":[{"check_id":"0..1","path":"semgrep-core/src/datalog/Datalog_experiment.ml","start":{"line":68,"col":1,"offset":2954},"end":{"line":89,"col":13,"offset":3505},"extra":{"message":"<internalonly>","metavars":{"$F":{"start":{"line":68,"col":5,"offset":2958},"end":{"line":68,"col":12,"offset":2965},"abstract_content":"dump_il","unique_id":{"type":"AST","md5sum":"d891ad6b9c0a591356b1bde9a603ebf5"}}}}},{"check_id":"0..1","path":"semgrep-core/src/engine/Eval_generic.ml","start":{"line":117,"col":1,"offset":4042},"end":{"line":127,"col":13,"offset":4373},"extra":{"message":"<internalonly>","metavars":{"$F":{"start":{"line":117,"col":5,"offset":4046},"end":{"line":117,"col":17,"offset":4058},"abstract_content":"print_result","unique_id":{"type":"AST","md5sum":"eff96a141279c318c34ccb1a8cd0176b"}}}}},{"check_id":"0..0.0.0","path":"semgrep-core/src/engine/Eval_generic.ml","start":{"line":119,"col":13,"offset":4096},"end":{"line":119,"col":22,"offset":4105},"extra":{"message":"<internalonly>","metavars":{}}},{"check_id":"0..0.0.0","path":"semgrep-core/src/engine/Eval_generic.ml","start":{"line":122,"col":19,"offset":4159},"end":{"line":122,"col":39,"offset":4179},"extra":{"message":"<internalonly>","metavars":{}}},{"check_id":"0..0.0.0","path":"semgrep-core/src/engine/Eval_generic.ml","start":{"line":124,"col":18,"offset":4268},"end":{"line":124,"col":42,"offset":4292},"extra":{"message":"<internalonly>","metavars":{}}},{"check_id":"0..0.0.0","path":"semgrep-core/src/engine/Eval_generic.ml","start":{"line":125,"col":18,"offset":4311},"end":{"line":125,"col":41,"offset":4334},"extra":{"message":"<internalonly>","metavars":{}}},{"check_id":"0..0.0.0","path":"semgrep-core/src/engine/Eval_generic.ml","start":{"line":126,"col":14,"offset":4349},"end":{"line":126,"col":23,"offset":4358},"extra":{"message":"<internalonly>","metavars":{}}}],"errors":[{"check_id":"ParseError","path":"semgrep-core/src/core/Pattern_match.ml","start":{"line":45,"col":22},"end":{"line":45,"col":24},"extra":{"message":"Syntax error","line":"  rule_id : rule_id; [@equal fun a b -> a.id = b.id]"}},{"check_id":"ParseError","path":"semgrep-core/src/core/Rule.ml","start":{"line":56,"col":18},"end":{"line":56,"col":20},"extra":{"message":"Syntax error","line":"  pstr : string; [@equal fun _ _ -> true]"}},{"check_id":"ParseError","path":"semgrep-core/src/core/ast/AST_generic.ml","start":{"line":382,"col":40},"end":{"line":382,"col":42},"extra":{"message":"Syntax error","line":"  id_constness : constness option ref; [@equal fun _a _b -> true]"}}],"stats":{"okfiles":233,"errorfiles":3}}
--- end semgrep-core stdout ---
--- semgrep-core stderr ---
Some constructs are not handled yet
CST was:
Original backtrace:
Raised at file "stdlib.ml", line 29, characters 22-33
Called from file "src/parsing/tree_sitter/Parse_ocaml_tree_sitter.ml", line 3295, characters 15-26
Called from file "src/parsing/tree_sitter/Parse_ocaml_tree_sitter.ml", line 3316, characters 9-37

Some constructs are not handled yet
CST was:
Original backtrace:
Raised at file "stdlib.ml", line 29, characters 22-33
Called from file "src/parsing/tree_sitter/Parse_ocaml_tree_sitter.ml", line 3295, characters 15-26
Called from file "src/parsing/tree_sitter/Parse_ocaml_tree_sitter.ml", line 3316, characters 9-37

Some constructs are not handled yet
CST was:
Original backtrace:
Raised at file "stdlib.ml", line 29, characters 22-33
Called from file "src/parsing/tree_sitter/Parse_ocaml_tree_sitter.ml", line 3295, characters 15-26
Called from file "src/parsing/tree_sitter/Parse_ocaml_tree_sitter.ml", line 3316, characters 9-37

--- end semgrep-core stderr ---
An error occurred while invoking the semgrep engine; please help us fix this by creating an issue at https://github.com/returntocorp/semgrep

Exited with code exit status 2
CircleCI received exit code 2
```

It looks like some test targets were reformatted by ocamlformat, which introduced syntax that our ocaml parser doesn't support. @aryx 

I just need to find out what exactly. Possibly also disable ocamlformat on test targets.

The error section of the json result is:
```
  "errors": [
    {
      "check_id": "ParseError",
      "path": "semgrep-core/src/core/Pattern_match.ml",
      "start": { "line": 45, "col": 22 },
      "end": { "line": 45, "col": 24 },
      "extra": {
        "message": "Syntax error",
        "line": "  rule_id : rule_id; [@equal fun a b -> a.id = b.id]"
      }
    },
    {
      "check_id": "ParseError",
      "path": "semgrep-core/src/core/Rule.ml",
      "start": { "line": 56, "col": 18 },
      "end": { "line": 56, "col": 20 },
      "extra": {
        "message": "Syntax error",
        "line": "  pstr : string; [@equal fun _ _ -> true]"
      }
    },
    {
      "check_id": "ParseError",
      "path": "semgrep-core/src/core/ast/AST_generic.ml",
      "start": { "line": 382, "col": 40 },
      "end": { "line": 382, "col": 42 },
      "extra": {
        "message": "Syntax error",
        "line":
          "  id_constness : constness option ref; [@equal fun _a _b -> true]"
      }
    }
  ],
  "stats": { "okfiles": 233, "errorfiles": 3 }
```

Very nice error messages. Looks like the parser doesn't like these ppx extensions?

Here's the reformatting that took place and causes a parse error:

image

File: scripts/oncall-email/oncall.ml [hunk 3]
```diff
@@ -66,106 +51,96 @@ let error j s =
...
   | J.Object xs ->
       let o = List.assoc fld xs in
       o
   | _ -> failwith (spf "could not find %s" fld)
 
-let jstring j =
-  match j with
-  | J.String s -> s
-  | _ -> error j "not a string"
+let jstring j = match j with J.String s -> s | _ -> error j "not a string"
```
Comment (IagoAbal) [inline scripts/oncall-email/oncall.ml]: This kind of reformatting doesn't look very nice...

Comment (IagoAbal) [reply scripts/oncall-email/oncall.ml]: Just to be more specific, I don't think that making as much code fit into an N-char line is the best formatting strategy.

**Comment by aryx** [comment]
Also we need to do those things in the weekend when everybody pushed their PRs.

**Comment by IagoAbal** [comment]
Overall I like the formatting applied by ocamlformat, and it will be nice not having to worry about that. I am just a little bit unhappy about ocamlformat trying to force things into a single line. It's a small thing, but if I'm not alone here then I believe it can be configured choosing a `sparse` style over the `compact` one.

**Comment by aryx** [comment]
I think my vote would go to non-compact too.

File: scripts/oncall-email/oncall.ml (hunk 3 above)
Comment (mjambon) [reply scripts/oncall-email/oncall.ml:61]: > Also we need to do those things in the weekend when everybody pushed their PRs.

yes, of course. This is a preview.

**Comment by mjambon** [comment]
> Also we need to do those things in the weekend when everybody pushed their PRs.

yes of course. This is meant as a preview.

File: scripts/oncall-email/oncall.ml (hunk 3 above)
Comment (mjambon) [reply scripts/oncall-email/oncall.ml]: Fixed with `break-cases = all` in the `.ocamlformat` file.

**Comment by mjambon** [comment]
I removed all actual reformatting from this branch and made other PRs to preview the different formatting styles:
* preview for this branch: https://github.com/returntocorp/semgrep/pull/3113/files
* preview for the `sparse` profile: https://github.com/returntocorp/semgrep/pull/3114/files (I think it's not what we want)

**Comment by mjambon** [comment]
I have various nitpicks on the indentation style but overall I think it's good enough. Keep in mind that so far we've been exceptionally aligned in terms of style. This protects us from future conflicts with contributors having different taste.

If you guys think that our code is now worse than before, then we could just drop ocamlformat and use only ocp-indent. I don't want to spend more time on tweaking this. For example, parens around anonymous functions don't align like I want but I'm willing to let it go.

**Comment by aryx** [comment]
The conventional mode looks good to me. I vote for that.
I think it improves the code sometimes, so that's good news.

**Comment by mjambon** [comment]
I left instructions on how to solve conflicts in the big reformat commit.

**Pull request merged**